### PR TITLE
Add generic vector function material

### LINKED ISSUE #18372: VectorFunctionMaterial
## Reason
Pronghorn uses vectors for the effective diffusivity. This type is missing everything for testing 

## Design
Add the vectorFunctionMaterial

## Impact
Help testing objects in Pronghorn, which will help one PR there, which will help for benchmarking PINSFV with SANA

### PR DESCRIPTION
**Pull request by GiudGiud**
Closes #18372

### CODE REVIEW

**Comment by GiudGiud** [comment]
Well this is an interesting failure. Looks like there's something wrong with threads here.

**Comment by GiudGiud** [comment]
ah the tensor maths is wrong is what is happening.

File: framework/src/materials/GenericFunctionVectorMaterial.C [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+  _num_props = num_names;
+
+  _properties.resize(num_names);
+
+  if (_enable_stateful)
+  {
+    _properties_old.resize(num_names);
+    _properties_older.resize(num_names);
+  }
```
Comment (GiudGiud) [inline framework/src/materials/GenericFunctionVectorMaterial.C:68]: this is the same code as in GenericFunctionMaterial and it s all there is

I wonder if we re missing some code there?

File: framework/include/materials/GenericFunctionVectorMaterial.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Material.h"
+
+// Forward Declarations
```
Comment (andrsd) [inline framework/include/materials/GenericFunctionVectorMaterial.h:14]: cpt. Obvious. - probably caused by copy-and-paste => remove this.

File: framework/include/materials/GenericFunctionVectorMaterial.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  std::vector<std::string> _prop_names;
+  std::vector<FunctionName> _prop_values;
+
+  unsigned int _num_props;
+
+  std::vector<GenericMaterialProperty<RealVectorValue, is_ad> *> _properties;
+  std::vector<const MaterialProperty<RealVectorValue> *> _properties_old;
+  std::vector<const MaterialProperty<RealVectorValue> *> _properties_older;
+  std::vector<const Function *> _functions;
```
Comment (andrsd) [inline framework/include/materials/GenericFunctionVectorMaterial.h:51]: Missing doxygen for all the above member variables.

File: framework/src/materials/GenericFunctionVectorMaterial.C [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+registerMooseObject("MooseApp", GenericFunctionVectorMaterial);
+registerMooseObject("MooseApp", ADGenericFunctionVectorMaterial);
+
+defineLegacyParams(GenericFunctionVectorMaterial);
+
+template <bool is_ad>
+InputParameters
+GenericFunctionVectorMaterialTempl<is_ad>::validParams()
+{
+
```
Comment (andrsd) [inline framework/src/materials/GenericFunctionVectorMaterial.C:22]: Remove this empty line

File: framework/src/materials/GenericFunctionVectorMaterial.C [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                                            "The names of the properties this material will have");
+  params.addParam<std::vector<FunctionName>>("prop_values",
+                                             "The corresponding names of the "
+                                             "functions that are going to provide "
+                                             "the values for the variables, "
+                                             "minor ordering by component");
+  params.addDeprecatedParam<bool>("enable_stateful",
+                                  false,
+                                  "Enable the declaration of old and older values",
+                                  "all properties can implicitly become stateful");
```
Comment (andrsd) [inline framework/src/materials/GenericFunctionVectorMaterial.C:36]: Probably don't want to add a deprecated parameter in a new object. Recommended to remove and patch the code below not to use it.

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by GiudGiud** [comment]
Should be good now. I applied all your comments to GenericFunctionMaterial as well.

File: framework/src/materials/GenericFunctionVectorMaterial.C (hunk 1 above)
Comment (GiudGiud) [reply framework/src/materials/GenericFunctionVectorMaterial.C]: it s deleted. no failures. So that was deprecated code.

**Review by andrsd — APPROVED** [review]

**Pull request merged**